workflows/cloudify_system_workflows/dsl_import_resolver/resolver_with_catalog_support.py

```diff
@@ -130,6 +141,83 @@ def _resolve_plugin_yaml_url(self, import_url):
         plugin = self._find_plugin(name, plugin_filters)
         return self._make_plugin_yaml_url(plugin)
 
+    @staticmethod
+    def _create_zip(source, destination, include_folder=True):
+        with closing(zipfile.ZipFile(destination, 'w')) as zip_file:
+            for root, _, files in os.walk(source):
+                for filename in files:
+                    file_path = os.path.join(root, filename)
+                    source_dir = os.path.dirname(source) if include_folder \
+                        else source
+                    zip_file.write(
+                        file_path, os.path.relpath(file_path, source_dir))
+        return destination
+
+    @staticmethod
+    def _download_file(url, dest_path, target_filename=None):
+        with requests.get(url, stream=True, timeout=(5, None)) as resp:
+            resp.raise_for_status()
+            if not target_filename:
+                target_filename = os.path.basename(url)
+            file_path = os.path.join(dest_path, target_filename)
+            with open(file_path, 'wb') as f:
+                for chunk in resp.iter_content(chunk_size=8192):
+                    if chunk:
+                        f.write(chunk)
+        return file_path
+
+    def _upload_missing_plugin(self, name, specifier_set, distribution):
+        plugins_data = urllib.request.urlopen(PLUGIN_CATALOG_URL).read()
+        plugins_data = json.loads(plugins_data)
+        requested_plugin_data = [x for x in plugins_data if x["name"] == name]
+
+        matching_versions = []
+        for plugin in requested_plugin_data:
+            p_icon = plugin['icon']
+            matching_versions.extend([
+                (v, parse_version(v)) for v in plugin['versions'].keys()
```

Comment:
I don't understand why the raw data we're pulling from has the same plugin repeated multiple times. If we could make the source data cleaner then this function could be cleaner too.